### PR TITLE
MINOR: Update JUnit to 4.13 and annotate log cleaner integration test

### PR DESCRIPTION
**Pull request by ijuma**
JUnit 4.13 fixes the issue where `Category` and `Parameterized` annotations
could not be used together. It also deprecates `ExpectedException` and
`assertThat`. Given this, we:
- Replace `ExpectedException` with the newly introduced `assertThrows`.
- Replace `Assert.assertThat` with `MatcherAssert.assertThat`.
- Annotate `AbstractLogCleanerIntegrationTest` with `IntegrationTest` category.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
https://github.com/apache/kafka/pull/3333 and https://github.com/apache/kafka/pull/3977 were two previous attempts at annotating the log cleaner integration test.

**Comment by ijuma** [comment]
Relevant sections of gradle's profile report that take more than 5 seconds:

> :core | 1m56.36s | (total)
> :core:compileScala | 47.057s |  
> :core:unitTest | 38.768s |  
> :core:compileTestScala | 29.112s
> 
> :clients | 1m18.65s | (total)
> :clients:unitTest | 1m10.78s |  
> 
> connect:runtime | 48.457s | (total)
> :connect:runtime:unitTest | 46.623s |  
> 
> :streams | 47.635s | (total)
> :streams:unitTest | 40.710s |  
> 
> tools | 30.031s | (total)
> :tools:unitTest | 28.552s
> 
> :streams:streams-scala | 16.430s | (total)
> :streams:streams-scala:unitTest | 12.939s |

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:839]: The use of `ExpectedException` seems like it was unnecessary here in the first place -- we could have just used the `expected` parameter of the `@Test` annotation, which seems already the much more common pattern given we don't need to further inspect the thrown exception in most cases.

In cases we don't need to further inspect the exception, the main value I see here is that it is more targeted -- it validates that the specific call you expect to throw the exception is the one that throws it. Do we want to standardize on this form moving forward? If so, we'd probably need to make committers aware since most are probably using the annotation-based approach for the most part.

Also, seems like this reveals that the test was broken (or at least not calling the final `consumer.close` previously?

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java]: why the shift to checking exact v2 rather than that it is >= if zstd? given our default would be to preserve the feature moving forward, it seems better to write these tests to work for future updates by default (as appears to be the case with the code previously) and only have to edit them if we made a change that wasn't compatible with v2 zstd support.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:467]: We probably don't need it tested as many times as this class was covering it, but the switch to `assume` seems like it loses coverage of this error case since assume just skips the test. Admittedly just a one line exception case, but still better to have a regression test.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java]: If we're starting to use `Assume`, this seems like it should be one since the condition just skips the whole test?

**Review by ewencp — COMMENTED** [review]
A couple of comments, but mostly looks good as it is mostly a mechanical change.

You got my hopes up with this, I've been refreshing the junit 4.13 release issue daily waiting for beta to be approved and released as its final version :)

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:839]: Yeah, we should move away from the annotation based approach as it was removed from JUnit 5 and the issue you mentioned (you often don't know if the exception was thrown from the right method). I was going to mention that as part of my JUnit 5 discussion.

I'll check the close issue.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java]: Just an oversight, I think I did it right in the other class. Will fix.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:467]: Yes, will fix.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java]: I was thinking the same, but didn't do it in the end. Will change.

**Comment by ijuma** [comment]
@ewencp I was refreshing it regularly too and eventually decided that we don't need to wait for it since the rate of change in that branch is nearly 0 and the beta is in the central Maven repo. We'll just bump to final when that happens.

**Comment by ijuma** [comment]
JDK 11 build passed, JDK 8 build had one flake:

> java.lang.AssertionError: 
> Expected: <[KeyValue(0, 0), KeyValue(0, 1), KeyValue(0, 3), KeyValue(0, 6), KeyValue(0, 10), KeyValue(0, 15), KeyValue(0, 21), KeyValue(0, 28), KeyValue(0, 36), KeyValue(0, 45), KeyValue(0, 55), KeyValue(0, 66), KeyValue(0, 78), KeyValue(0, 91), KeyValue(0, 105), KeyValue(0, 120), KeyValue(0, 136), KeyValue(0, 153), KeyValue(0, 171), KeyValue(0, 190)]>
>      but: was <[KeyValue(0, 0), KeyValue(0, 1), KeyValue(0, 3), KeyValue(0, 6), KeyValue(0, 10), KeyValue(0, 15), KeyValue(0, 21), KeyValue(0, 28), KeyValue(0, 36), KeyValue(0, 45), KeyValue(0, 10), KeyValue(0, 21), KeyValue(0, 33), KeyValue(0, 46), KeyValue(0, 60), KeyValue(0, 75), KeyValue(0, 91), KeyValue(0, 108), KeyValue(0, 126), KeyValue(0, 145)]>
> 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
> 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
> 	at org.apache.kafka.streams.integration.EosIntegrationTest.checkResultPerKey(EosIntegrationTest.java:212)
> 	at org.apache.kafka.streams.integration.EosIntegrationTest.shouldNotViolateEosIfOneTaskFailsWithState(EosIntegrationTest.java:440)

cc @guozhangwang @mjsax

**Comment by ijuma** [comment]
@ewencp I've addressed the comments and fixed a few more things I noticed.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java]: magic comparison to v2 has same equality change here as in the previous test. any reason not to `>=`?

**Review by ewencp — APPROVED** [review]
One last nit on the version checking in one of the tests, but otherwise LGTM (modulo getting a clean, non-timed-out build)

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (mumrah) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:839]: Oh wow, this is so much nicer than the rule thing or explicit catches. +1

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (mumrah) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:65]: Since the `build` method of the builder is what you're testing here, I don't think the Supplier is necessary.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java]: Sorry for missing this one, fixed.

**Review by mumrah — APPROVED** [review]
LGTM, one small comment inline

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:65]: The checks happen in the constructor, right?

```java
if (magic < RecordBatch.MAGIC_VALUE_V2) {
            if (isTransactional)
                throw new IllegalArgumentException("Transactional records are not supported for magic " + magic);
            if (isControlBatch)
                throw new IllegalArgumentException("Control records are not supported for magic " + magic);
            if (compressionType == CompressionType.ZSTD)
                throw new IllegalArgumentException("ZStandard compression is not supported for magic " + magic);
        }
```

**Comment by ijuma** [comment]
retest this please

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java
Comment (mumrah) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:65]: Gotcha. I (incorrectly) assumed the "build" method.

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
Tests passed, merging to trunk.

**Pull request merged**
